text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,488 @@
+> For example, achieving about a 1% false positive rate would take on the order
+> of about ~1k base64-encoded bytes. However, in order to extract a construct list
+> from this Bloom Filter would be a considerable amount of work: we would need
+> a full list of all construct available in every library and at every version,
+> in order to test each construct's presence in the filter. This imposes a lot
+> of complexity on the receiver, which is probably not worth the space savings.
+
+### Analytics payload
+
+The analytics payload itself will be preceded by an encoding scheme identifier,
+followed by a `:` and the transmitted data:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Analytics: <payload-encoding>:<payload>
+```
+
+The only valid payload encodings at the moment are:
+
+* `deflate64`: base64-encoded, gzipped data.
+* `plain`: plaintext data.
+
+The payload itself is a JSON object with `"constructs"` currently defined as the only
+valid key (but can be extended in the future). Values can be prefixed with
+an encoding scheme as well:
+
+```json
+{
+  "constructs": "[<value-encoding>:]<prefix-encoded construct list>"
+}
+```
+
+The only valid payload value encodings at the moment are:
+
+* `afx`: affix-encoded string list (see Appendix B).
+* `plain`: plaintext data.
+
+Encoding may change the type (the encoded value is a `string` but the decoded
+value is a `string[]`).
+
+> If a value does not start with > /^[a-z0-9]{,10}:/ it is considered to
+> be not encoded (the encoding is considered to be "plain").
+
+## Metadata resource decoding
+
+Decoding will happen on the receiving end before the data hits long-term storage,
+in order to make consuming it as trivial as possible. There is no need to be frugal
+with bits anymore once the data hits an AWS datacenter.
+
+# Extensions
+
+We will be able more information into additional fields of the metadata resource.
+Size of template, number of resources are obvious candidates.
+
+We will be able to support non-jsii languages by adding metadata-specific APIs
+to the CDK.
+
+The jsii RTTI mechanism can be extended in the future as well, though those extensions
+are out of scope of this RFC.
+
+If we want to change the encoding of the payload in any way, we can.
+
+If the payload grows too large and changing the encoding scheme won't be
+enough to reduce it, we can transport and reference the payload as an asset.
+Though, that would require us to gain read access to a customer's S3 bucket
+from our account, which has security implications; it would also require a
+customer to have a bootstrapped account for *every* deployment, even small
+ones (though we could lift *that* limitation by hosting bootstrap buckets on behalf
+of users).
+
+# Unresolved questions
+
+- How acceptable is the giant base64-encoded blob in the template? If we want
+  to cut down on it, what clever strategies can we come up with?
```

Comment:
I think a key thing is we want to make sure *customers* are able to decode this data locally with great ease, so they can independently check that we are not collecting anything they would not want to disclose.